staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go

```diff
@@ -330,54 +307,69 @@ func isIncompatibleServerError(err error) bool {
 	return err.(*errors.StatusError).Status().Code == http.StatusUnsupportedMediaType
 }
 
+// GetObjects returns a (possibly cached) version of all the objects to apply
+// as a slice of pointer to resource.Info. The resource.Info contains the object
+// and some other denormalized data. This function should not be called until
+// AFTER the "complete" and "validate" methods have been called to ensure that
+// the ApplyOptions is filled in and valid. Returns an error if the resource
+// builder returns an error retrieving the objects.
+func (o *ApplyOptions) GetObjects() ([]*resource.Info, error) {
+	if !o.objectsCached {
+		// include the uninitialized objects by default if --prune is true
+		// unless explicitly set --include-uninitialized=false
+		r := o.Builder.
+			Unstructured().
+			Schema(o.Validator).
+			ContinueOnError().
+			NamespaceParam(o.Namespace).DefaultNamespace().
+			FilenameParam(o.EnforceNamespace, &o.DeleteOptions.FilenameOptions).
+			LabelSelectorParam(o.Selector).
+			Flatten().
+			Do()
+		if err := r.Err(); err != nil {
+			return nil, err
+		}
+		infos, err := r.Infos()
+		if err != nil {
+			return nil, err
+		}
+		o.objects = infos
+		o.objectsCached = true
+	}
+	return o.objects, nil
+}
+
+// SetObjects stores the set of objects (as resource.Info) to be
+// subsequently applied.
+func (o *ApplyOptions) SetObjects(infos []*resource.Info) {
+	o.objects = infos
+	o.objectsCached = true
+}
+
 // Run executes the `apply` command.
 func (o *ApplyOptions) Run() error {
-	var openapiSchema openapi.Resources
-	if o.OpenAPIPatch {
-		openapiSchema = o.OpenAPISchema
-	}
 
 	dryRunVerifier := &DryRunVerifier{
 		Finder:        cmdutil.NewCRDFinder(cmdutil.CRDFromDynamic(o.DynamicClient)),
 		OpenAPIGetter: o.DiscoveryClient,
 	}
 
-	// include the uninitialized objects by default if --prune is true
-	// unless explicitly set --include-uninitialized=false
-	r := o.Builder.
-		Unstructured().
-		Schema(o.Validator).
-		ContinueOnError().
-		NamespaceParam(o.Namespace).DefaultNamespace().
-		FilenameParam(o.EnforceNamespace, &o.DeleteOptions.FilenameOptions).
-		LabelSelectorParam(o.Selector).
-		Flatten().
-		Do()
-	if err := r.Err(); err != nil {
+	infos, err := o.GetObjects()
```

Comment:
Does this mean we have to cache all the objects locally (rather than one at a time)? is this fine?

Reply:
Yes--that is correct. It should be fine because we already do it (look at the slice of "objs", which stores all runtime.Object applied, and which I've removed). This is necessary in order to PreProcess (e.g. sort) or PostProcess all objects being applied.

Reply:
sounds good, thanks!